### PR TITLE
Guard against re-entrance in PendingWriteQueue.removeAndWriteAll()

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

PendingWriteQueue should guard against re-entrant writes once removeAndWriteAll() is run.

Modifications:

Continue writing until queue is empty.

Result:

Correctly guard against re-entrance.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler PTAL ...

File: transport/src/main/java/io/netty/channel/PendingWriteQueue.java [hunk 1]
```diff
@@ -166,38 +166,35 @@ public void removeAndFail(Throwable cause) {
...
         assert ctx.executor().inEventLoop();
 
-        if (size == 1) {
-            // No need to use ChannelPromiseAggregator for this case.
-            return removeAndWrite();
-        }
-        PendingWrite write = head;
-        if (write == null) {
-            // empty so just return null
+        if (isEmpty()) {
```
Comment (nmittler) [inline transport/src/main/java/io/netty/channel/PendingWriteQueue.java]: Is there any concerned of the overhead of always using the promisecombiner? I suspect having only one write is fairly rare .. is that your thinking here?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/PendingWriteQueue.java]: yes... Also the problem is that we basically need to ensure we can add more writes during the loop.

**Comment by nmittler** [comment]
@normanmaurer LGTM

**Comment by Scottmitch** [comment]
@normanmaurer - The logic in this method is meant to mimic `removeAndFailAll` ... consider moving this method directly after `removeAndFailAll` so in the future if there are any changes it will be more clear that they both should be updated.

**Comment by normanmaurer** [comment]
+1

**Comment by normanmaurer** [comment]
@Scottmitch moved... ok to merge ?

**Comment by Scottmitch** [comment]
lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (6d70f4b) and 4.0 (7a75551)

**Pull request closed without merging**